engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java

```diff
@@ -785,7 +786,22 @@ public boolean wasCached() {
      * @return {@code true} when the type differs from the cached value.
      */
     public boolean isTypeChanged() {
-      return !Arrays.equals(types, cachedTypes);
+      String[] visibleType = null;
+      String[] hiddenType = null;
+      if (typeInfo != null) {
+        visibleType = typeInfo.visibleType();
+        hiddenType = typeInfo.hiddenType();
+      }
+
+      String[] cachedVisibleType = null;
+      String[] cachedHiddenType = null;
+      if (cachedTypeInfo != null) {
+        cachedVisibleType = cachedTypeInfo.visibleType();
+        cachedHiddenType = cachedTypeInfo.hiddenType();
+      }
+
+      return !Arrays.equals(visibleType, cachedVisibleType)
```

Comment:
Shouldn't this delegate to `typeInfo.equals(...)` method? Once properly implemented to deal with `String[]` or once we use `List<String>` instead?